Test with checker

IliaKash opened the pull request.

IliaKash committed:
Функция клонирования

IliaKash committed:
Cloning function

IliaKash committed:
Cloning Function v3

IliaKash committed:
Cloning function final

IliaKash committed:
Merge branch 'mvg-internship:main' into main

IliaKash committed:
Merge branch 'mvg-internship:main' into main

IliaKash committed:
Test with checker

ssmolov commented on src/gate/debugger/checker.cpp:
```diff
@@ -14,6 +14,37 @@
 namespace eda::gate::debugger {
 
+bool Checker::areEqual(const GNet &lhs,
+                const GNet &rhs,
```
По аналогии с Вашими правками в checker.h, лучше сделать единообразное выравнивание аргументов метода. Так:
bool Checker::areEqual(const GNet &lhs,
                                           const GNet &rhs,
...

ssmolov commented on src/gate/debugger/checker.h:
```diff
@@ -78,6 +79,10 @@ class Checker final {
+  
+  bool areEqual(const GNet &lhs,
+                const GNet &rhs,
+                eda::gate::premapper::PreMapper::GateIdMap &gmap) const;
```
Такой длинный идентификатор типа лучше спрятать в using-алиас, например:
using GateIdMap = eda::gate::premapper::PreMapper::GateIdMap;

ssmolov commented on src/gate/model/gnet.h:
```diff
@@ -506,6 +506,15 @@ class GNet final {
+  // Cloning 
+  //===--------------------------------------------------------------------===//
+
+  /// Clones the net
```
К какому методу относится этот комментарий? К обоим?

Такие "висящие" Doxygen-like комментарии лучше не оставлять "посреди нигде". В случае с методом clone() я вообще сомневаюсь, что нужен какой-то отдельный комментарий, кроме заголовочного. Метод "clone(std::unordered_map<Gate::Id, Gate::Id> &oldToNewId)" в этом смысле интереснее, потому что он принимает на вход какие-то данные. Если имеет смысл держать его публичным, то вот к нему комментарий точно хотелось бы иметь, чтобы понять, что такое oldToNewId.

ssmolov commented on test/gate/model/gnet_test.cpp:
```diff
@@ -267,3 +267,21 @@ TEST(GNetTest, GNetRandTestIssue11877) {
+TEST(GNetTest, GNetWithCheckerTest) {
+  eda::gate::debugger::Checker checker;
+  auto net = makeRand(7, 5);
+  std::unordered_map<Gate::Id, Gate::Id> TestMap = {};
```
Переменные у нас именуются в стиле lowerCamelCase (см. README, я там пополнил раздел Coding Convention)

ssmolov commented on test/gate/model/gnet_test.h:
```diff
@@ -9,6 +9,7 @@
 #pragma once
 
 #include "gate/model/gnet.h"
+#include "gate/debugger/checker.h"
```
В каждой категории include необходимо придерживаться алфавитного порядка их следования

ssmolov reviewed (commented):
Разобрались с пропажей clone из ветки main -- его стёрли случайно:) Виновный покаялся))

Прочие замечания см. ниже

IliaKash committed:
Test with checker v2

ssmolov commented on src/gate/debugger/checker.h:
```diff
@@ -29,6 +30,7 @@ class Checker final {
   using GateBinding = std::unordered_map<Gate::Link, Gate::Link>;
   using SubnetBinding = std::unordered_map<GNet::SubnetId, GNet::SubnetId>;
   using GateConnect = Context::GateConnect;
```
Для пущей аккуратности я бы и using-алиасы предложил упорядочить по алфавиту.

ssmolov commented on src/gate/model/gnet.cpp:
```diff
@@ -649,6 +649,52 @@ void GNet::sortTopologically() {
+// Cloning 
+//===--------------------------------------------------------------------===//
+
+/// Clones the net
```
В конце короткого комментария лучше поставить точку.

ssmolov commented on src/gate/debugger/checker.h:
```diff
@@ -78,6 +80,10 @@ class Checker final {
+  
+  bool areEqual(const GNet &lhs,
+                const GNet &rhs,
+                eda::gate::premapper::PreMapper::GateIdMap &gmap) const;
```
Раз алиас GateIdMap был введен, то и используйте его -- это сократит строку

ssmolov commented on src/gate/model/gnet.h:
```diff
@@ -506,6 +506,15 @@ class GNet final {
+
+  GNet* clone();
+
+  /// oldToNewId stores correspondence between gates of the original net and the clone
```
Лучше так:

/// The 'oldToNewId' stores ... clone one.

Проверьте, что строка не вылезает за 80 символов. Если вылезает -- надо переформулировать.
Вообще просто текстом указывать в комментариях имена переменных -- не самая хорошая практика. Их трудно отлавливать при переименованиях. Для таких вещей лучше использовать вещи типа "@param" в Doxygen-комментариях.

IliaKash committed:
Test with checker v3

IliaKash committed:
Test with checker v3.1

ssmolov commented on src/gate/model/gnet.h:
```diff
@@ -506,6 +506,18 @@ class GNet final {
+
+  GNet* clone();
+
+  /** Clones the net
```
В конце предложения поставить точку.

ssmolov commented on src/gate/model/gnet.h:
```diff
@@ -506,6 +506,18 @@ class GNet final {
+
+  /** Clones the net
+   *  @param oldToNewId correspondence between gates. Original net -> cloned
+   *  @return The clone
```
The clone of what? лучше написать the clone of this net -- раз выше Вы с с маленькой буквы начинаете описание аргумента метода.

ssmolov commented on src/gate/model/gnet.h:
```diff
@@ -506,6 +506,18 @@ class GNet final {
+  GNet* clone();
+
+  /** Clones the net
+   *  @param oldToNewId correspondence between gates. Original net -> cloned
```
В поле @param предполагаются краткие комментарии относительно смысла аргументов. Предлагаю либо объединить два предложения в одну фразу, либо пояснение "Original net -> cloned" перенести в основной блок текста -- он помещается между заголовочным комментарием и блоком "@param".

ssmolov reviewed (commented):
Ещё одно общее замечание -- старайтесь писать осмысленные сообщения к коммитам. При слиянии Вашей ветки с main/master я, конечно, уберу бесконечные "update" и "fix", но надо, чтобы хотя бы один коммит в этой ветке имел осмысленное описание. В данном случае это будет что-то вроде "gate/model: test for clone that uses eq checker"

IliaKash committed:
gate/model: test for clone that uses eq checker

ssmolov reviewed (commented):
Попытался собрать проект с данным pull request, получил следующие предупреждения:

utopia-eda/build/../src/gate/model/gnet.cpp:671:18: warning: Potential leak of memory pointed to by 'newGate' [clang-analyzer-cplusplus.NewDeleteLeaks]
      oldToNewId[gate->id()] = newGate->id();
                 ^
/home/smolov/utopia-eda/build/../src/gate/model/gnet.cpp:658:7: note: Assuming the condition is false
  if (_gates.empty()) {
      ^
/home/smolov/utopia-eda/build/../src/gate/model/gnet.cpp:658:3: note: Taking false branch
  if (_gates.empty()) {
  ^
/home/smolov/utopia-eda/build/../src/gate/model/gnet.cpp:662:10: note: Calling 'GNet::clone'
  return clone(oldToNewId);
         ^
/home/smolov/utopia-eda/build/../src/gate/model/gnet.cpp:667:7: note: Assuming the condition is false
  if (oldToNewId.empty()) {
      ^
/home/smolov/utopia-eda/build/../src/gate/model/gnet.cpp:667:3: note: Taking false branch
  if (oldToNewId.empty()) {
  ^
/home/smolov/utopia-eda/build/../src/gate/model/gnet.cpp:689:7: note: Assuming the condition is true
  if (!_subnets.empty()) {
      ^
/home/smolov/utopia-eda/build/../src/gate/model/gnet.cpp:689:3: note: Taking true branch
  if (!_subnets.empty()) {
  ^
/home/smolov/utopia-eda/build/../src/gate/model/gnet.cpp:691:28: note: Calling 'GNet::clone'
      resultNet->addSubnet(subnet->clone(oldToNewId));
                           ^
/home/smolov/utopia-eda/build/../src/gate/model/gnet.cpp:667:7: note: Assuming the condition is false
  if (oldToNewId.empty()) {
      ^
/home/smolov/utopia-eda/build/../src/gate/model/gnet.cpp:667:3: note: Taking false branch
  if (oldToNewId.empty()) {
  ^
/home/smolov/utopia-eda/build/../src/gate/model/gnet.cpp:689:7: note: Assuming the condition is true
  if (!_subnets.empty()) {
      ^
/home/smolov/utopia-eda/build/../src/gate/model/gnet.cpp:689:3: note: Taking true branch
  if (!_subnets.empty()) {
  ^
/home/smolov/utopia-eda/build/../src/gate/model/gnet.cpp:691:28: note: Calling 'GNet::clone'
      resultNet->addSubnet(subnet->clone(oldToNewId));
                           ^
/home/smolov/utopia-eda/build/../src/gate/model/gnet.cpp:667:7: note: Assuming the condition is false
  if (oldToNewId.empty()) {
      ^
/home/smolov/utopia-eda/build/../src/gate/model/gnet.cpp:667:3: note: Taking false branch
  if (oldToNewId.empty()) {
  ^
/home/smolov/utopia-eda/build/../src/gate/model/gnet.cpp:689:7: note: Assuming the condition is true
  if (!_subnets.empty()) {
      ^
/home/smolov/utopia-eda/build/../src/gate/model/gnet.cpp:689:3: note: Taking true branch
  if (!_subnets.empty()) {
  ^
/home/smolov/utopia-eda/build/../src/gate/model/gnet.cpp:691:28: note: Calling 'GNet::clone'
      resultNet->addSubnet(subnet->clone(oldToNewId));
                           ^
/home/smolov/utopia-eda/build/../src/gate/model/gnet.cpp:667:7: note: Assuming the condition is true
  if (oldToNewId.empty()) {
      ^
/home/smolov/utopia-eda/build/../src/gate/model/gnet.cpp:667:3: note: Taking true branch
  if (oldToNewId.empty()) {
  ^
/home/smolov/utopia-eda/build/../src/gate/model/gnet.cpp:670:23: note: Memory is allocated
      Gate *newGate = new Gate(gate->func(), newSignals);
                      ^
/home/smolov/utopia-eda/build/../src/gate/model/gnet.cpp:671:18: note: Potential leak of memory pointed to by 'newGate'
      oldToNewId[gate->id()] = newGate->id();
                 ^
[60/61] Building CXX object src/CMakeFiles/umain.dir/main.cpp.o
/home/smolov/utopia-eda/build/../src/main.cpp:129:9: warning: unused type alias 'Link' [clang-diagnostic-unused-local-typedef]
  using Link = RtlContext::Link;
        ^
/home/smolov/utopia-eda/build/../src/main.cpp:130:9: warning: unused type alias 'GateBinding' [clang-diagnostic-unused-local-typedef]
  using GateBinding = RtlContext::Checker::GateBinding;
        ^
../src/main.cpp: In function ‘bool check(RtlContext&)’:
../src/main.cpp:129:9: warning: typedef ‘using Link = using Link = using Link = struct eda::base::model::Link<unsigned int>’ locally defined but not used [-Wunused-local-typedefs]
  129 |   using Link = RtlContext::Link;
      |         ^~~~
../src/main.cpp:130:9: warning: typedef ‘using GateBinding = using GateBinding = class std::unordered_map<eda::base::model::Link<unsigned int>, eda::base::model::Link<unsigned int> >’ locally defined but not used [-Wunused-local-typedefs]
  130 |   using GateBinding = RtlContext::Checker::GateBinding;
  
  Нужно их изжить. Попробуйте самостоятельно решить эту проблему, если не получится -- сигнализируйте. Вижу, что некоторые предупреждения носят странный характер (к последнему это не относится), надо понять, почему они возникают.

ssmolov commented on test/gate/model/gnet_test.cpp:
```diff
@@ -267,3 +267,21 @@ TEST(GNetTest, GNetRandTestIssue11877) {
+  EXPECT_TRUE(net.get()->clone()->nEdges() == net.get()->nEdges());
+}
+
+TEST(GNetTest, GNetAdressTest) {
```
Переименовать в GNetAddressTest

ssmolov commented on src/gate/model/gnet.h:
```diff
@@ -506,6 +506,19 @@ class GNet final {
+  GNet* clone();
+
+  /** Clones the net.
+   *  Original net -> cloned.
```
Я, вероятно, догадаюсь, что здесь речь идёт о семантике аргумента метода, но если эту строку увидит кто-то другой, то возникнут вопросы. Не поленитесь и напишите предложение полностью ("The input map provides correspondence ...")

IliaKash committed:
gate/model: test for clone that uses eq checker review fixup

ssmolov committed:
gate/model: fix "clone" method for GNet

To avoid clang-tidy warnings some refactoring is made.
Some convenience methods are used (newGate, setGate).
Moreover, for parameterized "clone" method an assumption
is made, that input map is empty upon method calling.

Signed-off-by: Sergey Smolov <smolov@ispras.ru>

IliaKash committed:
Merge pull request #1 from ssmolov/fix.gnet.clone

gate/model: fix "clone" method for GNet
